### PR TITLE
feat: Timeweights

### PR DESCRIPTION
**Pull request by OpheliaMiralles**
## Description

Originally designed to be able to provide lead time decay weights for nowcasting, but can be useful for any time of weighting scheme in the time dimension.

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Partially solves #63 

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [ ] I have tested the changes on a single GPU
-   [ ] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [x] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [x] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [x] My code follows the style guidelines of this project
-   [x] I have updated the documentation and docstrings to reflect the changes
-   [x] I have added comments to my code, particularly in hard-to-understand areas

## Additional Notes

The relative_date_indices might overlap with #153 for some parts, given that it was originally inspired by the time interpolation branch published on anemoi-training before the refactoring.

### CODE REVIEW

**Review by jakob-schloer — COMMENTED** [review]
Nice idea and work! Now that the loss refactor is merged to main, I believe that the time weighting functionality should be a loss scalar.

**Comment by OpheliaMiralles** [comment]
> Nice idea and work! Now that the loss refactor is merged to main, I believe that the time weighting functionality should be a loss scaler. The `timeweights.py` should probably go under `losses/scalers`, and the instantiation and keywords should be under `training.scalers`. I think one could assume that the `dataloader.validation_rollout` keyword is the maximum rollout length that would allow us to not touch the datamodule and dataset. Happy to help with that!

Thanks! I was not sure the loss refactoring was over, I'll restart working on my open PRs soon.

File: training/src/anemoi/training/losses/scalers/timeweights.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+
+class LeadTimeDecayScaler(BaseScaler):
+    """Class to weight each specific predicted lead time in the batch."""
+
+    scale_dims = OutputTensorDim.TIME
+
+    def __init__(
+        self,
+        relative_date_indices: list[int],
```
Comment (jakob-schloer) [inline training/src/anemoi/training/losses/scalers/timeweights.py]: Could we use a different keyword here? My first thought was that the indices are related to actual dates in the dataset. I think `leadtime_indices` or something like that would better.

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 2]
```diff
@@ -56,6 +57,7 @@ def __init__(
         data_indices: IndexCollection,
         metadata: dict,
         supporting_arrays: dict,
+        relative_date_indices: Optional[dict] = None,
```
Comment (jakob-schloer) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: Can you briefly explain why you need the `relative_date_indices`? The LeadTimeScaler does not use them, aren't they

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 3]
```diff
@@ -179,6 +184,15 @@ def define_delayed_scalers(self) -> None:
...
             self.loss.update_scaler(scaler=self.scalers[name][1], name=name)
 
+    def define_time_varying_scaling(self, lead_time: int) -> None:
+        """Update time varying scalers such as lead time decay scaling."""
+        for name, scaler_builder in self.time_varying_scalers.items():
+            self.scalers[name] = scaler_builder.get_scaling_values(lead_time=lead_time)
+            if self.loss.scaler.has_scaler(name):
+                self.loss.update_scaler(scaler=self.scalers[name][1], name=name)
+            else:
+                self.loss.add_scaler(*self.scalers[name], name=name)
```
Comment (jakob-schloer) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: `add_scaler` requires a dimensionality to be passed. I'm a bit confused now, if your scaler has TensorDim or OutputTensorDim.

Comment (OpheliaMiralles) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: The idea would be that it does not have a dim at all (because it is the same number for the whole tensor for each lead time) and can be expanded to match the tensor's dims

File: training/src/anemoi/training/train/forecaster/forecaster.py (hunk 2 above)
Comment (OpheliaMiralles) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: Indeed I think we could remove

**Comment by OpheliaMiralles** [comment]
Should we close it until we have a clearer view of what to do? @HCookie

**Comment by HCookie** [comment]
If you do not have capacity to work on this, mark it is a draft and we can keep it for ideas for later revisiting.

File: training/src/anemoi/training/losses/scaler_tensor.py [hunk 4]
```diff
@@ -304,6 +304,23 @@ def __exit__(context_self, *a):  # noqa: N805
...
+            Name or index of the scaler to check
+
+        Returns
+        -------
+        bool
+            True if the scaler exists, False otherwise
+        """
+        if isinstance(name, int):
+            return name in self._specified_dimensions.values()
+        return name in self.tensors
```
Comment (OpheliaMiralles) [inline training/src/anemoi/training/losses/scaler_tensor.py]:
```suggestion
```

File: training/src/anemoi/training/losses/scaler_tensor.py [hunk 5]
```diff
@@ -304,6 +304,7 @@ def __exit__(context_self, *a):  # noqa: N805
 
         return FrozenStateRecord()
 
+
```
Comment (OpheliaMiralles) [inline training/src/anemoi/training/losses/scaler_tensor.py]:
```suggestion
```
